gh-100926: Move ctype's pointers cache to StgInfo (#131282)
python/cpython: Doc/library/ctypes.rst

   ValueError: NULL pointer access
   >>>

.. _ctypes-pointers-ctypes-like-types:

Pointers for ctypes-like types
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In some cases you want that :func:`POINTER` can accept non-ctypes types. To do so

> just add to your types a class level attribute with the name ``__pointer_type__``.

Reviewer: We normally do not document/teach internal manipulation.

Author: Yeah. I don't have a strong opinion on this. But we have to document `__pointer_type__` or add a function/class to make users class suitable to use with `POINTER`.

Reviewer: > add a function/class to make users class suitable to use with POINTER

Another PR please, if you mean `set_type()` like function called in the `POINTER()`.

Author: Yeah. I want wait Petr's decision about this before moving forward.

Reviewer 2: Please remove this section.

.. function:: POINTER(type, /)

   Create or return a ctypes pointer type. Pointer types are cached and
   reused internally, so calling this function repeatedly is cheap.
   *type* must be a ctypes-like type. The ctypes-like type is a type that

>    has class level attribute with the name ``__pointer_type__``.

Reviewer: I think it's not preferable to mention that.

Reviewer 2: - Let's not talk about “ctypes-like” types.
- I have no issue with mentioning where the cache is. However, I don't think you can cleanly use it.

```suggestion
   *type* must be a ctypes type.
   
   .. impl-detail::

      The resulting pointer type is cached in the ``__pointer_type__``
      attribute of *type*.
      It is possible to set this attribute before the first call to
      ``POINTER`` in order to set a custom pointer type.
      However, doing this is discouraged: manually creating a suitable
      pointer type is difficult without relying on implementation
      details that may change in future Python versions.
```

      library. *name* is the name of the symbol that exports the data, *library*
      is the loaded shared library.

   Common class variables of ctypes data types:

   .. attribute:: __pointer_type__

      This attributes is a pointer type that was created by calling

>       :func:`POINTER` for corresponding ctypes data type. If ``POINTER`` was
>       not called for this type then attribute contains ``None``.

Reviewer 2: ```suggestion
      :func:`POINTER` for corresponding ctypes data type. If a pointer type
      was not yet created, the attribute is missing.
```

      library. *name* is the name of the symbol that exports the data, *library*
      is the loaded shared library.

   Common class variables of ctypes data types:

   .. attribute:: __pointer_type__

>       This attributes is a pointer type that was created by calling

Reviewer 2: There's a typo (attribute<mark>s</mark>)... but it's in an unnecessary phrase.
```suggestion
      The pointer type that was created by calling
```

      valid.  This object is only exposed for debugging; never modify the
      contents of this dictionary.

   .. attribute:: __pointer_type__

      This attributes is a pointer type that was created by calling
      :func:`POINTER` for corresponding ctypes data type. If ``POINTER`` was

>       not called then this attribute contains ``None``.

Reviewer: This should be referred/linked from the warning. Users might want the docstring as well.  Requirements would be:

* This is an (internal?) cache for the `PONTER()`.

* This initially contains `None`.

* A pointer type is stored here (only once) when initialized with the ctypes data type.

cc @junkmd

Reviewer 3: I think we should state that this attribute is used internally or for metaprogramming, and does not need to be considered in normal use cases.

Also, since `__pointer_type__` is a class attribute, putting it in the "Common instance variables" section might be misleading.
(While it's not strictly wrong in Python since "everything is an object", it would be less ambiguous to call it a "class variable".)

Reviewer: Optional:

* Non-ctype types can immitate the attribute for `POINTER()`.

The proxy demonstrates this.
